### PR TITLE
Add mailto links to worker management.

### PR DESCRIPTION
**Pull request by vdbergh**
A draft message is provided (which includes the message in the worker management panel) which of course can be edited.

<hr>

Hi user00,

It seems your worker user00-7cores-1234567 has some issue(s):

Your worker is producing excessive timelosses! 

Can you take a look?

You may possibly find more information at
https://34.205.74.97/actions?text=%22user00-7cores-1234567%22.

Note: it may be that your worker is currently blocked. In that case
you can unblock it at https://34.205.74.97/workers/user00-7cores-1234567.

Thanks in advance!
The Fishtest team

### CODE REVIEW

**Comment by vdbergh** [comment]
It works well on my desktop, but on my cell phone the spaces in the draft are replaced by +'es.

**Comment by ppigazzini** [comment]
On my desktop, the "+" replaces all the spaces in the draft of Thunderbird v 115.2.0

**Comment by vdbergh** [comment]
Maybe solved now. + -> %20

**Comment by dav1312** [comment]
Will this show the email of users in dev?

**Comment by ppigazzini** [comment]
> Will this show the email of users in dev?

For the approver.

**Comment by vondele** [comment]
let's please be careful about these things. Also, in case sending email is automatic, this needs to be approached very very carefully. If emails are sent by mistake or too frequently that's a real issue. Similarly, if this is a no-reply email, I think that's potentially an issue.

**Comment by ppigazzini** [comment]
> let's please be careful about these things. Also, in case sending email is automatic, this needs to be approached very very carefully. If emails are sent by mistake or too frequently that's a real issue. Similarly, if this is a no-reply email, I think that's potentially an issue.

Try on DEV, there is a link that opens the email client with a draft message.

**Comment by vondele** [comment]
OK, that's good.

**Comment by vondele** [comment]
Should there be a checkbox or a date, that the approver could click if an email has been sent, to avoid repeated emails?

**Comment by vdbergh** [comment]
With the new version the only thing that does not work for me is gmail on ipad as handler of mailto links (the usual mail app works fine). In gmail all spaces in the draft are suppressed. 

Ipad is always a bit strange so maybe no big deal....

**Comment by vdbergh** [comment]
Hmm with the new version Thunderbird as a handler in firefox no longer shows the draft (blank body).

It would be nice to know how firefox actually calls thunderbird.

**Comment by vdbergh** [comment]
It seems that thunderbird simply ignores the body argument of a mailto link. E.g.
```
/usr/lib/thunderbird/thunderbird mailto:user00@user00.user00?body=test
```
give a blank body. That's quite shocking.

**Comment by vondele** [comment]
probably version dependent, since it works with ` Thunderbird 102.15.0`

**Comment by vdbergh** [comment]
That's the version also have, but it seems it was badly packaged (Ubuntu 22.04). Using command line arguments to compose a message also does not work.

**Comment by vondele** [comment]
Ok, on Ubuntu 20.04 here, so yeah could be, or could be some security concern?

**Comment by Disservin** [comment]
I dont think so, it's rather common to prefill a mail with some html template, but js cant be added and html alone is not a security concern

**Comment by vdbergh** [comment]
@vondele Should there be a checkbox or a date, that the approver could click if an email has been sent, to avoid repeated emails?

Perhaps but it requires some thought what is the correct UI. In principle a worker can have two issues with some time in between which would require sending emails with different messages.

**Comment by vondele** [comment]
yes, a checkbox / date was my initial idea, but maybe it indeed needs some more thinking.

**Review by ppigazzini — APPROVED** [review]
Looks good on DEV

File: server/fishtest/views.py [hunk 1]
```diff
@@ -139,14 +140,50 @@ def login(request):
     return {}
 
 
+def worker_email(worker_name, message, host_url, blocked):
+    owner = worker_name.split("-")[0]
+    wrapped_message = textwrap.fill(message, width=70)
+    body = f"""\
+Hi {owner},
+
+It seems your worker {worker_name} has some issue(s):
```
Comment (vondele) [inline server/fishtest/views.py:149]: Maybe something like:

Thank you for contributing to the development of Stockfish. Unfortunately, it seems your fishtest worker {worker_name} has some issue(s):

Comment (ppigazzini) [reply server/fishtest/views.py:149]: I write something like this:

Dear XXX,

  I'm YYY, one of the fishtest maintainers.

Thank you for contributing to the Stockfish/fishtest projects!

Unfortunately, your worker is having many time losses, see the Events log:

https://tests.stockfishchess.org/actions?action=crash_or_time&user=SimonDemel&text=

Feel free to contact me or by email or, better, on the Discord Stockfish server, #fishtest channel.

https://discord.com/invite/awnh2qZfTT

Enjoy your day,

**Comment by vdbergh** [comment]
@ppigazzini  I like your proposal but it needs to integrate the message in the panel (there may be other things than time losses).

So perhaps
```
Dear <owner>,

I'm <approver>, one of the fishtest maintainers.

Thank you for contributing to the development of Stockfish. Unfortunately, it seems your fishtest worker <worker_name> 
has some issue(s):

<message>

See also https://tests.stockfishchess.org/actions?text=<worker_name>

Feel free to reply to this email for more information, or else contact the Discord Stockfish server, #fishtest channel.

https://discord.com/invite/awnh2qZfTT

It may be that your worker is currently blocked which means that it will not play any games. In that case you can unblock it at https://tests.stockfishchess.org/workers/<worker_name> after resolving the issue(s).

Enjoy your day,

<approver>
```
Of course it is only a template. One still has a chance to modify the text before actually sending the email.


There is one inconvenient issue that we do not know the real names of the users. We could add them to the db for the people that want to. It makes sense to make it a requirement for approvers.

**Comment by ppigazzini** [comment]
Usually, I search for the real name on GitHub, in the email address, in the AUTHOR file etc.
When applicable, I report other similar cases, write some suggestions etc.
But these are steps to be made by the reporter case for case.

**Comment by vdbergh** [comment]
I added a commit with an alternative template.

**Comment by vdbergh** [comment]
> Usually, I search for the real name on GitHub, in the email address, in the AUTHOR file etc. When applicable, I report other similar cases, write some suggestions etc. But these are steps to be made by the reporter case for case.

Still it would make sense to add them to the db, at least for approvers. Their names are likely already in the commit log anway. So they are hardly a secret.

**Comment by ppigazzini** [comment]
Some contributors don't want to have their real name in AUTHORS file, use a real email in commits etc.

**Comment by vdbergh** [comment]
> Some contributors don't want to have their real name in AUTHORS file, or use a not incognito email in git commits etc.

Still many real names are public (e.g. from the author file). Why not add these to the db? Then we can use the real name if it is available and otherwise the username.

**Comment by vdbergh** [comment]
In principle the db is more private than the AUTHORs file.

File: server/fishtest/views.py [hunk 2]
```diff
@@ -139,22 +140,76 @@ def login(request):
     return {}
 
 
+def worker_email(worker_name, blocker_name, message, host_url, blocked):
+    owner = worker_name.split("-")[0]
+    wrapped_message = textwrap.fill(message, width=70)
+    body = f"""\
+Dear {owner},
+
+I'm {blocker_name}, one of the fishtest maintainers.
```
Comment (vondele) [inline server/fishtest/views.py:149]: probably approvers instead of maintainers ?

**Comment by ppigazzini** [comment]
DEV updated.
The emal link stop working with a blocking message with more than 259 characters, I wonder if there is a max length of the link.

**Comment by vdbergh** [comment]
gmail does not seem to have this problem. Neither does the mail app on ipad. What mail application are you using?

**Comment by ppigazzini** [comment]
I was using the combo Edge (chromium) + Thunderbid, with max supported length of the mailto link of 1810 characters.

This question hints to a Chromium limit:
https://stackoverflow.com/questions/4473959/effective-maximum-mailto-body-lengths

**Comment by vdbergh** [comment]
I tried Chromium on Ubuntu 22.04 and it opens the mailto links without problem (so does Chrome). They are less than 2000 characters (I estimate) so there should be no problem.

I wonder if it is not Thunderbird that is the problem. Can you open the mailto link directly with Thunderbird?
```sh
 thunderbird <mailto link>
```
I can test it myself since thunderbird does not work at all for me.

**Comment by ppigazzini** [comment]
I tried on a VM with the Windows Mail client.
The second link provided states that it's a long-time known problem with chromium on Windows.

**Comment by vdbergh** [comment]
The mailto link is indeed 1810 characters.  Which is less than 2000 (which according to the 2nd link is when the problems start). But you say there is a problem anyway. Does it also occur with Chrome? Chromium is not really a consumer browser.

Perhaps we can make the text area smaller.

**Comment by ppigazzini** [comment]
Chrome, Edge, Opera are chromium based and they don't work with a link bigger than 1810 characters.

**Comment by vdbergh** [comment]
On Linux they do...

Is 1810 the hard limit on Windows? Then enforcing the message to be smaller will fix the problem.

**Comment by ppigazzini** [comment]
I found the max working length of the mailto link adding and removing a character at time. However, the StackOverflow post reports a max length for the **encoded** link, and in the code the threshold is set to 1995, not 2000.

**Comment by vdbergh** [comment]
By coincidence the technologov link is exactly 1810 characters (with encoding).

**Comment by vdbergh** [comment]
I now cut the message off at 400 characters. Of course one needs to resubmit the form, otherwise the message will not change on the db.

We need to be smarter than just cutting the message off but perhaps not this PR.

**Comment by ppigazzini** [comment]
> By coincidence the technologov link is exactly 1810 characters (with encoding).

Yes, by coincidence :D

DEV updated.

**Comment by vdbergh** [comment]
Now a message is printed if the description is truncated.

**Comment by ppigazzini** [comment]
DEV updated.

**Comment by ppigazzini** [comment]
The Issue text message (with the technologov worker) must be max 256 chars to work on Windows with a chromium based browser. We should rework the emal content to save some of the other 1554 chars.

**Comment by vdbergh** [comment]
Hmm. So the restriction is much more severe than 1810 characters. This is weird since on Linux there seems to be no restriction and it is the same code base.

We could include a link to the message instead of including the message directly in the email.

Alternatively we could try to make the email shorter.

**Comment by vdbergh** [comment]
Ok I misunderstood. The length of 1810 of the technologov link was no accident but corresponded to a message of 256 characters (I thought the message was 500 characters).

That’s a serious problem. Perhaps dropping the last sentence (about unblocking yourself) might help.

**Comment by vdbergh** [comment]
In the latest commit I made the email shorter.

**Comment by ppigazzini** [comment]
DEV updated. With the new email template, an Issue message of 400 chars results in a mailto link of 1370 chars. Dropping the ***** helped a lot to save chars.

**Comment by ppigazzini** [comment]
PROD updated, thank you @vdbergh :)

I raised the max_chars clamp to 500 in https://github.com/official-stockfish/fishtest/pull/1808, the technologov mailto link after that is 1503 chars, so ve have ~ 200 chars for a future revision of the mail text.

**Pull request merged**